--- a/openquake/__init__.py
+++ b/openquake/__init__.py
@@ -49,8 +49,6 @@
     along with this program.  If not, see <http://www.gnu.org/licenses/>.
 """
 
-import os
-
 # Please note: the release date should always have a value of 0 (zero) in the
 # master branch. It will only be set to a meaningful value in *packaged* and
 # released OpenQuake code.
@@ -59,6 +57,3 @@
     4,  # minor
     6,  # sprint number
     0)  # release date (seconds since the "Epoch"), do *not* set in master!
-
-# The path to the OpenQuake root directory
-OPENQUAKE_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))
