--- a/bin/cache_gc.py
+++ b/bin/cache_gc.py
@@ -31,9 +31,6 @@
 import getopt
 import sys
 
-import oqpath
-oqpath.set_oq_path()
-
 from openquake import kvs
 from openquake import logs
 
--- a/bin/openquake
+++ b/bin/openquake
@@ -67,9 +67,6 @@
 except ImportError:
     sys.path.append('/usr/openquake')
 
-import oqpath
-oqpath.set_oq_path()
-
 from openquake import flags
 from openquake import __version__
 
--- a/bin/openquake_supervisor
+++ b/bin/openquake_supervisor
@@ -25,9 +25,6 @@
 import os
 import sys
 
-import oqpath
-oqpath.set_oq_path()
-
 
 def main():  # pylint: disable=C0111
     os.environ['DJANGO_SETTINGS_MODULE'] = 'openquake.settings'
