--- a/openquake/calculators/hazard/event_based/core.py
+++ b/openquake/calculators/hazard/event_based/core.py
@@ -109,7 +109,7 @@
             for rupture in ses[event_set]:
 
                 if self.calc_proxy['GMF_OUTPUT']:
-                    common_path = os.path.join(self.base_path,
+                    common_path = os.path.join(self.calc_proxy.base_path,
                             self.calc_proxy['OUTPUT_DIR'],
                             "gmf-%s-%s" % (str(event_set.replace("!", "_")),
                                            str(rupture.replace("!", "_"))))
