Index: nhlib/nhlib/bin/check_gsim.py
===================================================================
--- nhlib.orig/nhlib/bin/check_gsim.py	2012-09-13 08:49:19.143930171 +0200
+++ nhlib/nhlib/bin/check_gsim.py	2012-09-13 08:50:43.447927851 +0200
@@ -261,7 +261,7 @@
     return sctx, rctx, dctx, stddev_types, expected_results, result_type
 
 
-if __name__ == '__main__':
+def nain():
     import argparse
 
     def gsim_by_import_path(import_path):
@@ -324,3 +324,7 @@
         print >> sys.stderr, stats
     if errors:
         exit(127)
+
+
+if __name__ == '__main__':
+    main()
Index: nhlib/setup.py
===================================================================
--- nhlib.orig/setup.py	2012-09-13 08:49:18.871930180 +0200
+++ nhlib/setup.py	2012-09-13 08:50:43.479927851 +0200
@@ -48,6 +48,10 @@
 
 
 setup(
+    entry_points = {
+        'console_scripts': [
+            'check_gsim = nhlib.bin.check_gsim:main',
+            ]},
     name='nhlib',
     version=version,
     description="nhlib is a library for performing seismic hazard analysis",
@@ -61,9 +65,8 @@
     ],
     ext_modules=[geodetic_speedups, geoutils_speedups],
     include_dirs=include_dirs,
-    scripts=['tests/gsim/check_gsim.py'],
-    maintainer='Anton Gritsay',
-    maintainer_email='anton@openquake.org',
+    maintainer='OpenQuake Team',
+    maintainer_email='devops@openquake.org',
     classifiers=(
         'Development Status :: 3 - Alpha',
         'Intended Audience :: Education',
Index: nhlib/nhlib/bin/__init__.py
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ nhlib/nhlib/bin/__init__.py	2012-09-13 08:52:16.251925298 +0200
@@ -0,0 +1,18 @@
+# nhlib: A New Hazard Library
+# Copyright (C) 2012 GEM Foundation
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+nhlib stands for New Hazard Library.
+"""
